sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/schema/complex/AvroArraySchema.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Override
+    public void add() {
+        this.state.push(this);
+        /* Read the block size, call onBlockCount. */
+        AvroLongSchema blockCountSchema = new AvroLongSchema(
+            this.state,
+            this::onBlockCount
+        );
+        blockCountSchema.add();
+    }
+
+    /**
+     * Block count handler
+     *
+     * @param blockCount The number of elements in the block.
+     */
+    private void onBlockCount(Long blockCount) {
+        /* If blockCount = 0, then we're done.*/
+        if (blockCount == 0) {
+            this.done = true;
+        /* If blockCount > 0, read the item, call onItem. */
+        } else if (blockCount > 0) {
+            this.blockCount = blockCount;
+            AvroSchema itemSchema = getSchema(
+                this.itemType,
+                this.state,
+                this::onItem
+            );
+            itemSchema.add();
+        /* If blockCount < 0, use absolute value, read the byteCount, call onByteCount. */
+        } else {
+            this.blockCount = -blockCount;
+            AvroLongSchema byteCountSchema = new AvroLongSchema(
+                this.state,
+                this::onByteCount
+            );
+            byteCountSchema.add();
+        }
+    }
+
+    /**
+     * Byte count handler.
+     *
+     * @param ignore The number of bytes in the block.
+     */
+    private void onByteCount(Long ignore) {
+        /* TODO (gapra) : Use this in case we want to skip through the array in the future. Not required for now. */
+        /* Read the item, call onItem. */
+        AvroSchema itemSchema = getSchema(
+            this.itemType,
+            this.state,
+            this::onItem
+        );
+        itemSchema.add();
+    }
+
+    /**
+     * Item handler.
+     *
+     * @param item The item.
+     */
+    private void onItem(Object item) {
+        /* Add the item to the list. */
+        this.result.add(item);
+
+        /* Decrement the block count. */
+        this.blockCount--;
+
+        /* If blockCount = 0, there are no more items in the block, read another blockCount and call onBlockCount. */
+        if (blockCount == 0) {
+            AvroLongSchema blockCountSchema = new AvroLongSchema(
+                this.state,
+                this::onBlockCount
+            );
+            blockCountSchema.add();
+        /* If blockCount != 0, there are more items in the block, read another item and call onItem. */
+        } else {
+            AvroSchema itemSchema = getSchema(
+                this.itemType,
+                this.state,
+                this::onItem
+            );
+            itemSchema.add();
+        }
+    }
+
+    @Override
+    public void progress() {
+        /* Progress is defined by progress on the sub-type schemas. */
+    }
+
+    @Override
+    public boolean canProgress() {
+        /* Can always make progress since it is defined by the progress on the sub-type schemas. */
+        return true;
+    }
```

Comment:
I think these are effectively not needed for complex types (like we discussed the other day). However, I'd rather not leave them no-op like that - in case we have bug in top level logic this could lead to infinite loop.

I'd rather make them throw OR remove from this interface. Consider such options:
- create PrivmitiveSchema interface where these are defined and not define them for complex types at all (this would require casting/instanceof when popping stack)
- create ComplexSchema abstract class and define them there in one place. Also make them throw some not supported exception so that they break loop in case there's some bug elsewhere.. (not ideal contract).

Reply:
The cleanest way to do this would be to just throw from complex types if isDone is false.

Reply:
added to issues https://github.com/Azure/azure-sdk-for-java/issues/10463

Reply:
@gapra-msft . Yet other option to consider:
In "complex" types we execute logic to listen to next part of that type in onXXX() listener methods. Maybe we should reduce their responsibility and push that part of logic into "progress" and "canProgress". So that instead of "returning" from child element we'd use "progress" to trigger listening to next subelement. Then "progress" would have meaning here. Additionally it would decrease complexity and resposibility of onXXX handlers with is a good thing to do.